fsspec/implementations/tests/test_prefix.py

```diff
@@ -0,0 +1,135 @@
+    ".test.fakedata.2.csv": (b"a,b\n" b"3,4\n"),
+}
+odir = os.getcwd()
+
+
+@contextmanager
+def filetexts(d, open=open, mode="t"):
+    """Dumps a number of textfiles to disk
+
+    d - dict
+        a mapping from filename to text like {'a.csv': '1,1\n2,2'}
+
+    Since this is meant for use in tests, this context manager will
+    automatically switch to a temporary current directory, to avoid
+    race conditions when running tests in parallel.
+    """
+    dirname = tempfile.mkdtemp()
+    try:
+        os.chdir(dirname)
+        for filename, text in d.items():
+            f = open(filename, "w" + mode)
+            try:
+                f.write(text)
+            finally:
+                try:
+                    f.close()
+                except AttributeError:
+                    pass
+
+        yield list(d)
+
+        for filename in d:
+            if os.path.exists(filename):
+                try:
+                    os.remove(filename)
+                except (IOError, OSError):
+                    pass
+    finally:
+        os.chdir(odir)
+
+
+def test_cats():
+    with filetexts(csv_files, mode="b"):
+        fs = PrefixFileSystem(prefix=".", filesystem=fsspec.filesystem("file"))
+        assert fs.cat(".test.fakedata.1.csv") == b"a,b\n" b"1,2\n"
+        out = set(fs.cat([".test.fakedata.1.csv", ".test.fakedata.2.csv"]).values())
+        assert out == {b"a,b\n" b"1,2\n", b"a,b\n" b"3,4\n"}
+        assert fs.cat(".test.fakedata.1.csv", None, None) == b"a,b\n" b"1,2\n"
+        assert fs.cat(".test.fakedata.1.csv", start=1, end=6) == b"a,b\n" b"1,2\n"[1:6]
+        assert fs.cat(".test.fakedata.1.csv", start=-1) == b"a,b\n" b"1,2\n"[-1:]
+        assert (
+            fs.cat(".test.fakedata.1.csv", start=1, end=-2) == b"a,b\n" b"1,2\n"[1:-2]
+        )
+        out = set(
+            fs.cat(
+                [".test.fakedata.1.csv", ".test.fakedata.2.csv"], start=1, end=-1
+            ).values()
+        )
+        assert out == {b"a,b\n" b"1,2\n"[1:-1], b"a,b\n" b"3,4\n"[1:-1]}
+
+
+def test_not_found():
+    fn = "not-a-file"
+    fs = PrefixFileSystem(prefix=".", filesystem=fsspec.filesystem("file"))
+    with pytest.raises((FileNotFoundError, OSError)):
+        with OpenFile(fs, fn, mode="rb"):
+            pass
+
+
+def test_isfile():
+    fs = PrefixFileSystem(prefix=".", filesystem=fsspec.filesystem("file"))
+    with filetexts(files, mode="b"):
+        for f in files.keys():
+            assert fs.isfile(f)
+            assert fs.isfile("file://" + f)
+        assert not fs.isfile("not-a-file")
+        assert not fs.isfile("file://not-a-file")
+
+
+def test_isdir():
+    fs = PrefixFileSystem(prefix=".", filesystem=fsspec.filesystem("file"))
+    with filetexts(files, mode="b"):
+        for f in files.keys():
+            assert fs.isdir(os.path.dirname(os.path.abspath(f)))
+            assert not fs.isdir(f)
+        assert not fs.isdir("not-a-dir")
+
+
+@pytest.mark.parametrize("prefix", ["/", "/tmp"])
```

Comment:
This needs to run on windows too, so we can't make assumptions about root/tmp dirs.

Reply:
Ok I refactored this test.

In general, I would like to re-use the tests already implemented in other file systems, wrapping it with the PrefixFileSystem. Do you think it is feasible? 
Or maybe it is better to just wait for https://github.com/intake/filesystem_spec/pull/651

Reply:
Reusing sounds great. IMPO, there is no need to test prefixfs with every single fs implementation, as that would be excessive. Since all filesystems follow the same interface and prefixfs is meant to be generic and doesn't make any fs-specific assumptions, your approach with only testing with localfs looks good to me.

Reply:
Ok makes sense! 

I fear that also the tests that use the localfs are not enough though, it would be nice to reuse the entire `LocalFileSystem` test suite for this `PrefixFileSystem`, because re-implementing them means duplicating a lot of code. But probably it is not possible right now...